File: solution/py/checkInclusion.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        if len(s1) > len(s2): return False
        
        s1Count, s2Count = [0] * 26, [0] * 26
        count = 0

        for i in range(len(s1)):
            s1Count[ord(s1[i]) - ord('a')] =+1
            s2Count[ord(s2[i]) - ord('a')] =+1

        for i in range(26):
            count += (1 if s1Count[i] == s2Count[i] else 0)

        lo = 0
        for hi in range(len(s1), len(s2)):

            if count == 26: return True
            
            index = ord(s2[hi]) - ord('a')
            s2Count[index]+= 1
            if s1Count[index] == s2Count[index]:
                count+=1
            elif s1Count[index] + 1 == s2Count[index]:
                count-=1

            index = ord(s2[lo]) - ord('a')
            s2Count[index]-= 1
            if s1Count[index] == s2Count[index]:
                count+=1
            elif s1Count[index] - 1 == s2Count[index]:
                count-=1

            lo+=1

        return count == 26

        

File: solution/java/checkInclusion.java

class checkInclusion {

    public static void main(String[] args) {
        System.out.println(checkInclusion("ab", "eidbaooo"));
    }

    public static boolean checkInclusion(String s1, String s2) {
        int[] s1Count = new int[26];
        int[] s2Count = new int[26];
        int count = 0;

        for (int i = 0; i < s1.length(); i++) {
            s1Count[s1.charAt(i) - 'a']++;
            s2Count[s2.charAt(i) - 'a']++;
        }

        for (int i = 0; i < 26; i++) {
            count += s1Count[i] == s2Count[i] ? 1 : 0;
        }

        for (int index = 0, lo = 0, hi = s1.length(); hi < s2.length(); hi++) {
            if (count == 26)
                return true;

            index = s2.charAt(hi) - 'a';
            s2Count[index]++;
            if (s1Count[index] == s2Count[index]) {
                count++;

            } else if (s1Count[index] + 1 == s2Count[index]) {
                count--;
            }

            index = s2.charAt(lo) - 'a';
            s2Count[index]--;
            if (s1Count[index] == s2Count[index]) {
                count++;

            } else if (s1Count[index] - 1 == s2Count[index]) {
                count--;
            }
            lo++;
        }

        return count == 26;
    }
}